servers/user-list-search/src/corpus/CorpusSearchQueryBuilder.ts

```diff
@@ -0,0 +1,414 @@
+   */
+  public static async fromQueryString(
+    opts: QuerySearchCorpusArgs,
+    keywordFallback = true,
+  ): Promise<SemanticSearchQueryBuilder | SimpleQueryStringBuilder> {
+    const query = await SemanticSearchQueryBuilder.getQueryVec(
+      opts.search.query,
+    );
+    if (query == null && keywordFallback) {
+      return new SimpleQueryStringBuilder(opts);
+    }
+    return new SemanticSearchQueryBuilder({ ...opts, search: { query } });
+  }
+
+  /**
+   * Return a JSON-like object which can
+   * be used as a body in ES/opensearch client requests
+   * See https://opensearch.org/docs/latest/search-plugins/knn/index/
+   */
+  public toJSON() {
+    return this.knnQuery();
+  }
+
+  /**
+   * Retrieve query vector from Sagemaker endpoint
+   * Public because mocking the internals for testing
+   * is a PITA and it's way easier to wrap the function itself.
+   * @returns the vector query for knn search, or undefined
+   * if it could not be retrieved.
+   */
+  public static async getQueryVec(
+    query: string,
+  ): Promise<Array<number> | undefined> {
+    const client = sagemakerClient();
+    const input: InvokeEndpointInput = {
+      EndpointName: embeddingsConfig.endpoint,
+      ContentType: 'application/json',
+      Body: Buffer.from(JSON.stringify({ inputs: query })),
+    };
+    Sentry.addBreadcrumb({
+      data: { query: query },
+    });
+    const command = new InvokeEndpointCommand(input);
+    try {
+      const response = await client.send(command);
+      const vector = JSON.parse(response.Body.transformToString());
+      // Probably should never be null/malformed without error, but ensure
+      if (vector?.[0]?.[0]?.length > 0) {
+        // Return [CLS] token
+        return vector[0][0];
+      }
+      return undefined;
+    } catch (error) {
+      Sentry.captureException(error);
+      return undefined;
+    }
+  }
+  /**
+   * Builds KNN query object for opensearch, with post-request
+   * filters.
+   * Filtering in the request requires an engine other than
+   * nmslib, which is currently used. Because of this there may
+   * be fewer results than `k` nearest neighbors.
+   */
+  private knnQuery() {
+    const k =
+      this.opts.pagination?.['limit'] ??
+      this.opts.pagination?.['first'] ??
+      config.pagination.defaultPageSize;
+    // esb doesn't support hybrid or knn queries
+    // for opensearch, since the APIs diverge
+    // For now, just do this semi-manually
+    const filter = this.filter();
+    return {
+      query: {
+        bool: {
+          filter: filter.toJSON(),
+          must: [
+            {
+              knn: {
+                [embeddingsConfig.propertyName]: {
+                  vector: this.opts.search.query,
+                  k,
+                },
+              },
+            },
+          ],
+        },
+      },
+    };
+  }
+}
+
+export class SimpleQueryStringBuilder extends SearchQueryFilters {
+  private builder: esb.RequestBodySearch;
```

Comment:
this is a renamed/moved version of the old keyword class without any changes except for inheritance and constructor